### ISSUE TITLE
Best approach to connecting WSGI Microsoft Teams Bot 

### ISSUE THREAD

**Issue by sphet** [buildbot/buildbot#5268]
I have a rudimentary status notifier working, but because the Microsoft Teams web-hook has limitations, creating threaded and transitional messages are not possible. When the status of a build changes over time I have to send multiple messages and they get out of order.  As a result I want to implement a chat bot that can post responses to the initial build and keep updating the status.

I have seen some python WSGI implementations of simple bots and now wonder if I can't write one that hooks into BuildBot. My question is how do I bridge from the notifier report system to a flask application running under twisted?

**Comment by tardyp** [comment]
Hi @sphet I am not sure exactly what is your design here. Looks like you have a separate server which runs your bot.
Buildbot can directly work as a chat bot. we already have a lot of reporters which does that. 
Doing it in async can be frightening, but it is not that hard especially for chats, which are already very event based :)

@macdems contributed lately a telegram bot, which is pretty cool and implements state chat you are looking forward.

https://github.com/buildbot/buildbot/pull/5016

The implementation is mostly independent from Telegram, so I guess you could make a team backend for word,py framework.

**Comment by sphet** [comment]
@tardyp I'll take a look at the telegram bot. It looks more full featured than what I want, but could be a good starting place for implementations.

**Comment by sphet** [comment]
Here is an update in case people are curious.

Microsoft Teams brokers all communication between team members and a bot through the main  Microsoft/Azure servers. The Microsoft BotBuilder SDK (for python) uses asiohttp as the async http engine. Microsoft bots can be self hosted, or hosted in the Azure platform. Their SDK is quite feature full, and there are some complications to getting it up and running in an unsolicited push-style setup, but I have it working. However, I do not run it as part of Build Bot because Microsoft requires a public https endpoint, and we have a many master to one Teams setup to handle. So I ended up writing a stand-alone Microsoft conversation bot that publishes a public https end point which is tunnelled through our firewall. It also offers a second set of end points on a different port only available inside our company network. I wrote a push-style notification based on the github sample which pushes the state changes of the builders to my internal chat bot endpoint. The chatbot is then responsible for tracking the builds and generating conversation threads with status updates and at-mentions to people who have broken the build.

Sadly, it is all bespoke and unfortunately proprietary, but I wanted to share my strategy here for others.
